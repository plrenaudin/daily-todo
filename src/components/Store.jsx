import React from "react";
import {
  LOAD,
  ADD,
  addTodo,
  REMOVE,
  removeTodo,
  listTodo,
  TOGGLE,
  toggleTodo,
  listWeekTodo
} from "../modules/model";
import { isoFormatDate, currentWeekISO } from "../modules/date";

const ConfigStore = React.createContext();
const TodoStore = React.createContext();
const WeekStore = React.createContext();

const todoReducer = (state, action) => {
  switch (action.type) {
    case LOAD:
      return action.data;
    case ADD:
      return addTodo(state, action.data);
    case REMOVE:
      return removeTodo(state, action.data);
    default:
      throw new Error();
  }
};

const weekReducer = (state, action) => {
  switch (action.type) {
    case LOAD:
      return action.data;
    case TOGGLE:
      return toggleTodo(state, action.todo, action.day);
    default:
      throw new Error();
  }
};

const useDayTodos = () => {
  const [week, dispatch] = React.useReducer(weekReducer, []);
  return { week, dispatch };
};

const useTodos = () => {
  const [todos, dispatch] = React.useReducer(todoReducer, []);
  return { todos, dispatch };
};

const Provider = ({ children }) => {
  //Config
  const [currenDate, setCurrentDate] = React.useState(isoFormatDate());
  const configStore = { currenDate, setCurrentDate };

  //Week
  const weekStore = useDayTodos();

  //Todos
  const todoStore = useTodos();

  React.useEffect(() => {
    const week = currentWeekISO();
    Promise.all([listTodo(), listWeekTodo(week)]).then(([todos, weekTodos]) => {
      todoStore.dispatch({ type: LOAD, data: todos });
      weekStore.dispatch({ type: LOAD, data: weekTodos });
    });
  }, []);

  return (
    <ConfigStore.Provider value={configStore}>
      <TodoStore.Provider value={todoStore}>
        <WeekStore.Provider value={weekStore}>{children}</WeekStore.Provider>
      </TodoStore.Provider>
    </ConfigStore.Provider>
  );
};

export { ConfigStore, WeekStore, TodoStore, Provider };
