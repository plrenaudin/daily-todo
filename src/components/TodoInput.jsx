import React from "react";
import { ADD } from "../modules/model";
import { TodoStore } from "../components/Store";

const TodoInput = React.memo(() => {
  let textInput = React.useRef();
  const createTodo = () => {
    if (!textInput.current.value) return;
    dispatch({ type: ADD, data: textInput.current.value });
    textInput.current.value = "";
  };
  const { dispatch } = React.useContext(TodoStore);
  return (
    <>
      <input
        type="text"
        ref={textInput}
        onKeyDown={e => e.keyCode === 13 && createTodo()}
      />
      <button onClick={createTodo}>Add</button>
    </>
  );
});

export { TodoInput };
