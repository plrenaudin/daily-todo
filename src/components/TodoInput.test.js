import React from "react";
import renderer from "react-test-renderer";
import { TodoInput } from "./TodoInput";
import { TodoStore } from "./Store";

it("renders the TodoInput correctly", () => {
  const component = (
    <TodoStore.Provider value={{ todos: ["plip", "huhu"], dispatch: () => {} }}>
      <TodoInput />
    </TodoStore.Provider>
  );
  expect(renderer.create(component).toJSON()).toMatchSnapshot();
});
