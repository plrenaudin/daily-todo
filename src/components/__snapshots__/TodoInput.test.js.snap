// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the TodoInput correctly 1`] = `
.c0 {
  padding: 0.5rem 0 0.5rem 0.3rem;
}

<input
  className="c0"
  onKeyDown={[Function]}
  placeholder="Add Todo"
  type="text"
/>
`;
